File: frontend/src/utils/pricing.ts
export type Accommodation = "double" | "single" | "none";

export const ACCOMMODATION_PRICE: Record<Accommodation, number> = {
  double: 179,
  single: 219,
  none: 0,
};

export const ACCOMMODATION_LABEL: Record<Accommodation, string> = {
  double: "Dvojlôžková izba",
  single: "Jednolôžková izba",
  none: "Bez ubytovania a stravy",
};

export const ACCOMMODATION_NOTE: Record<Accommodation, string> = {
  double: "2× nocľah, 2× raňajky, 2× obed, 2× večera, miestna daň",
  single: "2× nocľah, 2× raňajky, 2× obed, 2× večera, miestna daň",
  none: "Účasť len na prednáškach, bez ubytovania a stravy",
};

export const ACCOMMODATION_ORDER: Accommodation[] = ["double", "single", "none"];

/** A ZTP card holder pays no local tax, so their stay is 3 € cheaper. */
export const ZTP_DISCOUNT = 3;

export const ZTP_LABEL = "Vlastník preukazu ZŤP";

export const ZTP_HINT =
  `Držiteľ preukazu ZŤP neplatí miestnu daň — cena za pobyt sa znižuje ` +
  `o ${ZTP_DISCOUNT} €.`;

/** Price of one person's package, after the ZTP exemption from the local tax. */
export function personPrice(
  accommodation: Accommodation,
  ztp = false,
): number {
  const base = ACCOMMODATION_PRICE[accommodation];
  // Nothing is booked with "none", so there is no local tax to waive either.
  return ztp && accommodation !== "none" ? base - ZTP_DISCOUNT : base;
}

/** Room types that qualify for the employer recreation voucher (2-night stay). */
export function qualifiesForVoucher(accommodation: Accommodation): boolean {
  return accommodation !== "none";
}

export interface PricePerson {
  name: string;
  surname: string;
  accommodation: Accommodation;
  /** Holder of a ZTP card — exempt from the local tax. */
  ztp?: boolean;
}

export interface PriceLineItem {
  name: string;
  accommodation: Accommodation;
  price: number;
}

export interface PriceBreakdown {
  items: PriceLineItem[];
  /** Sum of the accommodation packages, without the voluntary contribution. */
  accommodationTotal: number;
  extraContribution: number;
  /** Full price of the stay plus the contribution, whoever ends up collecting it. */
  total: number;
  /** Settled at the hotel reception — the stay, when the recreation voucher is claimed. */
  paidAtHotel: number;
  /** Transferred to EVS: the contribution alone once the voucher moves the stay. */
  amountDue: number;
}

export function calculatePrice(
  people: PricePerson[],
  extraContribution = 0,
  recreationVoucher = false,
): PriceBreakdown {
  const items: PriceLineItem[] = people.map((p) => ({
    name: `${p.name} ${p.surname}`.trim(),
    accommodation: p.accommodation,
    price: personPrice(p.accommodation, p.ztp),
  }));

  const accommodationTotal = items.reduce((sum, item) => sum + item.price, 0);
  const extra = Math.max(0, Math.round(extraContribution) || 0);
  const paidAtHotel = recreationVoucher ? accommodationTotal : 0;

  return {
    items,
    accommodationTotal,
    extraContribution: extra,
    total: accommodationTotal + extra,
    paidAtHotel,
    amountDue: accommodationTotal + extra - paidAtHotel,
  };
}

File: backend/app/services/pricing.py
"""Package prices and what a registration actually owes.

Mirrors `frontend/src/utils/pricing.ts` — the two must stay in sync.
"""

# Price per attending person:
#   double – twin room, single – single room, none – lectures only.
ACCOMMODATION_PRICE = {"double": 179, "single": 219, "none": 0}

# A ZTP card holder is exempt from the local tax, so their stay costs 3 € less.
ZTP_DISCOUNT = 3


def person_price(person: dict) -> int:
    """Price of one person's package, after the ZTP exemption from the local tax."""
    accommodation = person.get("accommodation")
    base = ACCOMMODATION_PRICE.get(accommodation, 0)
    # Nothing is booked with "none", so there is no local tax to waive either.
    if person.get("ztp") and accommodation != "none":
        return base - ZTP_DISCOUNT
    return base


def voucher_claimed(doc: dict) -> bool:
    """Top-level voucher flag, falling back to the legacy per-person flags."""
    if "recreation_voucher" in doc:
        return bool(doc["recreation_voucher"])
    if doc.get("registrant", {}).get("recreation_voucher"):
        return True
    return any(a.get("recreation_voucher") for a in doc.get("attendees", []))


def accommodation_total(registrant: dict, attendees: list[dict]) -> int:
    """Price of the packages alone, without the voluntary contribution."""
    total = 0
    if registrant.get("is_attendee") and registrant.get("accommodation"):
        total += person_price(registrant)
    for a in attendees:
        total += person_price(a)
    return total


def hotel_amount(registrant: dict, attendees: list[dict], recreation_voucher: bool) -> int:
    """What is settled at the hotel reception — the stay, when a voucher is claimed."""
    return accommodation_total(registrant, attendees) if recreation_voucher else 0


def transfer_amount(
    registrant: dict,
    attendees: list[dict],
    extra_contribution: int = 0,
    recreation_voucher: bool = False,
) -> int:
    """Amount the registrant transfers to EVS.

    With a recreation voucher the stay is settled at the hotel reception, so only
    the voluntary contribution is invoiced — nothing at all without one.
    """
    stay = 0 if recreation_voucher else accommodation_total(registrant, attendees)
    return stay + max(0, int(extra_contribution or 0))


def amount_due(doc: dict) -> int:
    """What is left to transfer for a stored registration.

    An amount already sent in the payment e-mail wins over the calculation.
    """
    stored = doc.get("payment_amount")
    if stored is not None:
        return int(stored)
    return transfer_amount(
        doc["registrant"],
        doc.get("attendees", []),
        doc.get("extra_contribution", 0),
        voucher_claimed(doc),
    )
